Document using authd with NFS and Samba (#672)
ubuntu/authd: docs/howto/use-with-samba.md

# Using authd with Samba

> The UIDs and GIDs assigned to users and groups by authd are unique to each

Reviewer: ```suggestion
The user identifiers (UIDs) and group identifiers (GIDs) assigned by authd are unique to each
```

machine. This means that when using authd with Samba, the UIDs and GIDs of users
and groups on the Samba server will not match those on the client machines,
which leads to permission issues.

To avoid these issues, you can use Samba with ID mapping. This ensures that the

> UIDs and GIDs of users and groups are mapped correctly across all machines.

Reviewer: ```suggestion
UIDs and GIDs are mapped correctly across all machines.
```

> ## Setting up Samba with ID Mapping

Reviewer: ```suggestion
## Setting up Samba with ID mapping
```

> This guide will walk you through setting up a Samba server with ID mapping. The
> user `alice` will be able to access a shared directory on the server from a
> client machine.

Reviewer: ```suggestion
This guide will walk you through setting up a Samba server with ID mapping. By following
the steps outlined below, a user `alice` will be able to access a shared directory on the server from a
client machine.
```

---

> ### Steps for the Server

Reviewer: ```suggestion
### Configure the server
```

1. **Install Samba:**
   Install the Samba server package:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo apt update
   sudo apt install samba
   ```

> 2. **Create the Shared Directory:**

Reviewer: ```suggestion
2. **Create the shared directory:**
```

   Create the directory to be shared and set ownership to the `alice` user:

>    ```bash

Reviewer: ```suggestion
   ```text
```

Remove syntax highlighting for terminal commands.
It's not very helpful for bash and is not always consistent with what a user sees in their terminal.

   sudo mkdir -p /srv/samba/alice
   sudo chown alice:alice /srv/samba/alice
   ```

> 3. **Edit Samba Configuration:**

Reviewer: ```suggestion
3. **Edit Samba configuration:**
```

   Open the Samba configuration file:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo editor /etc/samba/smb.conf
   ```

   Add the following section at the end of the file:

   ```ini
   [alice]
   path = /srv/samba/alice
   browsable = yes
   writable = yes
   valid users = alice
   ```

>    - **Explanation:** This section defines a Samba share named `alice` located
>    at `/srv/samba/alice`. It is visible to users on the network (`browsable`),
>    allows writing (`writable`), and restricts access to the `alice` user (`valid
>    users`).

Reviewer: ```suggestion
<replace-with-triple-backticks>{admonition} Explanation
:class: information
This section defines a Samba share named `alice` located
at `/srv/samba/alice`. It is visible to users on the network (`browsable`),
allows writing (`writable`), and restricts access to the `alice` user (`valid
users`).
<replace-with-triple-backticks>
```

Author: Nice!

> 4. **Create a Samba User for `alice`:**

Reviewer: ```suggestion
4. **Create a Samba user for `alice`:**
```

   Add the `alice` user to the Samba database and set a password:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo smbpasswd -a alice
   ```

   Follow the prompts to set the Samba password for the user.

> 5. **Restart Samba Service:**

Reviewer: ```suggestion
5. **Restart Samba service:**
```

   Restart the Samba service to apply the changes:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo systemctl restart smbd
   ```

---

> ### Steps for the Client

Reviewer: ```suggestion
### Steps for the client
```

> 1. **Install Samba Client:**

Reviewer: ```suggestion
1. **Install Samba client:**
```

   Install the required packages for connecting to Samba shares:

   ```bash
   sudo apt update
   sudo apt install smbclient cifs-utils
   ```

> 2. **Test Access to the Share:**

Reviewer: ```suggestion
2. **Test access to the share:**
```

>    Test connectivity using `smbclient` (replace `$SERVER` with the Samba
>    server's hostname or IP address):

Reviewer: ```suggestion
   Test connectivity using `smbclient`, making sure to replace `$SERVER` with the Samba
   server's hostname or IP address:
```

Avoid parentheticals where possible.

   ```bash
   smbclient //$SERVER/alice -U alice
   ```

   Enter the Samba password for `alice` when prompted. If successful, a `smb: \>`
   prompt appears.

> 3. **Mount the Share:**

Reviewer: ```suggestion
3. **Mount the share:**
```

   Create a mount point for the share:

   ```bash
   mkdir -p /home/alice/samba
   ```

   Mount the share using the `cifs` filesystem type:

   ```bash
   sudo mount -t cifs //$SERVER/alice /home/alice/samba -o user=alice,uid=$(id -u alice),gid=$(id -g alice)
   ```

   Enter the Samba password for `alice` when prompted.

> 4. **Optional: Add to `/etc/fstab` for Persistent Mounting:**

Reviewer: ```suggestion
4. **Optional: Add to `/etc/fstab` for persistent mounting:**
```

Title could be more explicit: add _what_ to `/etc/fstab`?

Author: Changed to 

```
**Optional: Add the share to `/etc/fstab` for persistent mounting:**
```

   To automatically mount the share at boot, use a credentials file:

   - Create a credentials file:

     ```bash
     sudo editor /etc/samba/credentials
     ```

     Add the following lines:

     ```
     username=alice
     password=YOUR_PASSWORD
     ```

   - Secure the credentials file:

     ```bash
     sudo chmod 600 /etc/samba/credentials
     ```

   - Update `/etc/fstab`:

     ```
     //$SERVER/alice /home/alice/samba cifs credentials=/etc/samba/credentials,uid=alice,gid=alice 0 0
     ```

> 5. **Verify the Mount:**

Reviewer: ```suggestion
5. **Verify the mount:**
```

   As the user `alice`, try accessing the shared directory:

   ```bash
   ls -la /home/alice/samba
   ```

   Verify write access by creating a test file:

   ```bash
   touch /home/alice/samba/test
   ```

>    **Security Note:** Files and directories in the share may appear as owned by
>    `alice` on the client, but access control is enforced by the server. For
>    example:
> 
>    - If `alice` does not have permission on the server, access will be denied
>      even if ownership appears correct on the client.

Reviewer: ```suggestion
   <replace-with-triple-backticks>{admonition} Security note
   :class: note
   **Security Note:** Files and directories in the share may appear as owned by
   `alice` on the client, but access control is enforced by the server.
   
   For example, if `alice` does not have permission on the server, access will be
   denied even if ownership appears correct on the client.
   <replace-with-triple-backticks>
```

>    Example:

Reviewer: This seems like an optional step to verify a security feature.

Could "Example:" be replaced with a more descriptive step "6. **Test enforced access control (optional)**"

   - Create a restricted directory on the server:

     ```bash
     sudo mkdir /srv/samba/alice/secrets
     sudo chmod 700 /srv/samba/alice/secrets
     ```

   - Attempt to access it on the client:

     ```bash
     ls /home/alice/samba/secrets
     ```

>      **Result:** *Permission denied.*

Reviewer: Is "permission denied" the actual terminal output?

If so, it could be put in a codeblock, i.e.:

"The terminal output will indicate that the user does not have access to the restricted directory:

<replace-with-triple-backticks>text
Permission denied
<replace-with-triple-backticks>

Author: done

---

### Cleanup

> #### On the Server

Reviewer: ```suggestion
#### On the server
```

> 1. **Delete the Shared Directory:**

Reviewer: ```suggestion
1. **Delete the shared directory:**
```

   Remove the directory used for the Samba share:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo rm -rf /srv/samba/alice
   ```

> 2. **Purge Installed Samba Packages:**

Reviewer: ```suggestion
2. **Purge installed Samba packages:**
```

   If Samba is no longer needed, uninstall it completely:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo apt purge samba samba-common
   sudo apt autoremove
   ```

---

#### On the Client

> 1. **Unmount the Shared Directory:**

Reviewer: ```suggestion
1. **Unmount the shared directory:**
```

>    Unmount the shared directory:

Reviewer: This line, and some others below, merely repeat the message in the heading.
The text should give additional context, where possible.
If no additional context can be given, then consider removing the text because it is redundant.

Author: removed

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo umount /home/alice/samba
   ```

> 2. **Delete the Mount Point:**

Reviewer: ```suggestion
2. **Delete the mount point:**
```

>    Remove the mount point directory:

Reviewer: redundant?

>    ```bash

Reviewer: ```suggestion
   ```text
```

   rmdir /home/alice/samba
   ```

> 3. **Remove Entry from `/etc/fstab`:**

Reviewer: ```suggestion
3. **Remove fstab entry:**
```

   If you added the share to `/etc/fstab`, remove its entry:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo editor /etc/fstab
   ```

>    Locate and delete the line referencing the Samba share, then save and close.

Reviewer: If the user has followed this guide closely, we should be able to specify exactly the line that they need to locate and delete?

Author: How would you specify that? Like this?

 Locate and delete the Samba share entry:

   ```
   //$SERVER/alice /home/alice/samba cifs credentials=/etc/samba/credentials,uid=alice,gid=alice 0 0
   ```

Reviewer: Let's not add it: it's probably obvious to someone following the guide.

> 4. **Delete Credentials File:**

Reviewer: ```suggestion
4. **Delete credentials file:**
```

   If a credentials file was used, remove it:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo rm /etc/samba/credentials
   ```

> 5. **Purge Installed Samba Client Packages:**

Reviewer: ```suggestion
5. **Purge installed Samba client packages:**
```

   If Samba client tools are no longer needed, uninstall them:

>    ```bash

Reviewer: ```suggestion
   ```text
```

   sudo apt purge samba-common smbclient cifs-utils

>    sudo apt autoremove

Reviewer: autoremove is mentioned earlier too.

could we mention this once at the end and say why it is being used?

Author: Well it's once on the server and once on the client. It removes any packages that were installed as dependencies of the other packages.

Reviewer: Ah, of course! Point taken.
